pkg/image/prune/controller.go

```diff
@@ -0,0 +1,185 @@
+package prune
+
+import (
+	"fmt"
+	"strings"
+
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	kclient "github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/golang/glog"
+	gonum "github.com/gonum/graph"
+	"github.com/openshift/origin/pkg/api/graph"
+	"github.com/openshift/origin/pkg/client"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+type PruningController struct {
+	images                 client.ImageInterface
+	streams                client.ImageStreamInterface
+	pods                   kclient.PodInterface
+	replicationControllers kclient.ReplicationControllerInterface
+	registryURL            string
+}
+
+func (c *PruningController) Run() error {
+	images, err := c.images.List(labels.Everything(), fields.Everything())
+	if err != nil {
+		return fmt.Errorf("error listing all images: %v", err)
+	}
+
+	streams, err := c.streams.List(labels.Everything(), fields.Everything())
+	if err != nil {
+		return fmt.Errorf("error listing all image streams: %v", err)
+	}
+
+	pods, err := c.pods.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("error listing all pods: %v", err)
+	}
+
+	rcs, err := c.replicationControllers.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("error listing all replication controllers: %v", err)
+	}
+
+	g := graph.New()
+	c.addImagesToGraph(g, images)
+	c.addImageStreamsToGraph(g, streams)
+	c.addPodsToGraph(g, pods)
+	c.addReplicationControllersToGraph(g, rcs)
+	c.prune(g)
+
+	return nil
+}
+
+const minimumImageAgeInMinutesToPrune = 60
+
+func (c *PruningController) addImagesToGraph(g graph.Graph, images *imageapi.ImageList) {
```

Comment:
----- Original Message -----
> > +		return fmt.Errorf("error listing all replication controllers: %v", err)
> > +	}
> > +
> > +	g := graph.New()
> > +	c.addImagesToGraph(g, images)
> > +	c.addImageStreamsToGraph(g, streams)
> > +	c.addPodsToGraph(g, pods)
> > +	c.addReplicationControllersToGraph(g, rcs)
> > +	c.prune(g)
> > +
> > +	return nil
> > +}
> > +
> > +const minimumImageAgeInMinutesToPrune = 60
> > +
> > +func (c *PruningController) addImagesToGraph(g graph.Graph, images
> > *imageapi.ImageList) {
> 
> so would you do this?
> 
> ```go
> type ImagePruner struct {
>   g graph.Graph
>   images client.ImageInterface
>   // ...
> }
> ```
> 
> And what would the func that takes an Image that you pass to Run do?
> 

I don't think the Pruner needs to know about client.ImageInterface - the Function you pass to run would do.  That way you can dry run without having to change the algorithm.

So a set of stages:

1. Need to load and fetch objects and add them to a graph
2. Identify any images that aren't referenced by the graph
3. Delete those images

But you could replace #3 with print the images out.  Or replace #3 with a function that deletes the images AND removes them from an IR.  Or removes them from the IR, and then deletes them later.

It's important to keep that last step separate from the first two.

> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1841/files#r28806542